numpy/lib/function_base.py

```diff
@@ -1752,10 +1762,73 @@ def diff(a, n=1, axis=-1):
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
     if n > 1:
-        return diff(a[slice1]-a[slice2], n-1, axis=axis)
-    else:
+        return diff(a[slice1]-a[slice2], n-1, axis=axis, 
+            to_begin=to_begin, to_end=to_end)
+
+    # if nothing to add to either side, fast track the result
+    elif to_begin is None and to_end is None:
         return a[slice1]-a[slice2]
 
+    else:
+        # compute the length of the diff'd portion
+        # force length to be non negative
+        l_diff = a.shape[axis] - 1
+        if l_diff < 0:
+            l_diff = 0
+
+        # make to_end a 1D array
+        if to_end is None:
+            l_end = 0
+        else:
+            to_end = np.atleast_1d(to_end)
+            if to_end.ndim == 1:
+                l_end = len(to_end)
+            else:
+                to_end = np.array(to_end, ndmin=nd)
+                l_end = to_end.shape[axis]
+
+        # make to_begin a 1D array
+        if to_begin is None:
+            l_begin = 0
+        elif isinstance(to_begin, str) and to_begin == 'first':
+            to_begin = a.take([0], axis)
+            l_begin = 1
+        else:
+            to_begin = np.atleast_1d(to_begin)
+            if to_begin.ndim == 1:
+                l_begin = len(to_begin)
+            else:
+                to_begin = np.array(to_begin, ndmin=nd)
+                l_begin = to_begin.shape[axis]
+
+        # initialize the output array with correct shape
+        shape = list(a.shape)
+        shape[axis] = l_begin + l_end + l_diff
+        result = np.empty(tuple(shape), dtype=a.dtype)
+
+        # wrap ndarray subclasses
+        wrap = getattr(a, "__array_prepare__", a.__array_wrap__)
```

Comment:
Ha! Mine were https://github.com/numpy/numpy/blob/master/numpy/core/fromnumeric.py#L42 and https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L4608
Some grepping suggests `__array_wrap__` is used more often than `__array_prepare__`, but there is no obvious conclusion. 

The documentation is not terribly clear -- see  https://docs.scipy.org/doc/numpy-1.11.0/reference/arrays.classes.html -- but it does suggest that by default `__array_prepare__` does nothing while `__array_wrap__` changes the class to be that of the instance to which it is attached. I think that makes slightly more sense here. Since it also simplifies the code, I would suggest just to use `a.__array_wrap__` unconditionally.